Improve react print

carlobeltrame opened the pull request:
Lots of improvements and fixes in react print. Will continue to add commits until I have reviews.

Before: [before.pdf](https://github.com/ecamp/ecamp3/files/9714555/before.pdf)
After: [current.pdf](https://github.com/ecamp/ecamp3/files/9714492/current.pdf)


### Fixes
- [x] Fix visual bugs in picasso and schedule entry rendering 973add77d0fca3163f7dfac026e8f656150ee3a1 3f546c93f8de516bb8cfd172e9bf0e87c557d154 67f445f94108a64c44f81a0326ff488d52e7ee93 613e38134c10c24f1da8e57105a6e67ed6531da1 c81948e11934b39b7e55c7a977292bd99e7aa244
- [x] Import components directly from react-pdf, now that they are exported correctly there, and eliminate our adapter middle man file 97721adb39c2366c8aa1ff55ec01ad8925226b8e
- [x] Various refactoring for better reusability of code a658ec5bd753be0d8c9282fa5aaa248d5ec82397 2c249b98c8c0c002befbf00c046084d7bae555e7 8eb9bf478e95ed14fa5bffc6687f2285853548e3 ad989d8cefdda9717618607964332ece53dca10f
- [x] Fix the last known remaining issues with web workers (apart from bad reloading support in Vite) e8b8bb3348d0b63ad4b808d2d0506140cc45f895
- [ ] Replace the deprecated vite-plugin-worker to enable the upgrade to Vite 3

### Feature parity with nuxt print
- [x] Storyboard first and last column is treated as plain text 6d0e7e7cc2e60b8dff421456388b6408ccc3b9e4
- [x] Ordered list enumeration ca7147cc83f049f16ae88140aec501ed67bf29a2
- [x] Use Vuetify algorithm for schedule entry stacking (now we could really remove the db columns `left` and `width` on the ScheduleEntry entity) 9abf776add244783a26b184b9e8567e47e1277c3
- [x] Table of contents f816d62c5a80f4510d231b52705a922a038f4508
- [x] Cover page ce291957c9cede5764f7817b915bb5ae5642f50a
- [x] Story 35b8e8ce103d4799bd25621024f6af47fa0e24b1
- [x] Page numbers
- [x] Chunked picasso
- [ ] Content node icons
 
### New features
- [x] Display dayResponsibles on picasso c905b2e6204ca6cf3d8bc4ca4503f2dd5e36e8cf
- [x] Schedule entries on the picasso now have adaptive vertical padding, to account for very short schedule entries 3293002bf387c27dd68fbdf24129bee6907a0e15
- [x] Include PDF bookmarks in the document, but hide the bookmarks sidebar by default when displaying the PDF e74729b23fe814872cd33c1f4c5dd00cbe2527b9
- [x] Automatically select the relevant part of the day to display on the picasso
- [ ] Day overview
- [ ] E2e tests for PDF (using cypress, and possibly https://github.com/jeetiss/react-pdf-testing-library)
- [ ] Use white font on dark schedule entry backgrounds

### Experimental / trying out
- [x] Create a [custom renderer](https://lachlan-miller.me/articles/vue-3-pdf-customer-renderer) for [Vue 3](https://vuejs.org/api/custom-renderer.html) which can take in a Vue component structure and output react-pdf's internal layouting structure, and generate a pdf using react-pdf but without react
  - :arrow_right: Current problem is that custom renderers are only available in Vue 3, and we can't compile a Vue 3 component structure with vite 2.

carlobeltrame committed:
Fix picasso rendering

carlobeltrame committed:
Return to importing directly from react-pdf

carlobeltrame committed:
Extract stylesheets into separate file

This will make theming and customization easier in the future

carlobeltrame committed:
Display dayResponsibles on picasso

carlobeltrame committed:
Fix vanishing text on short schedule entries

This involved setting the time columns to a carved-in-stone constant
width. Otherwise, as soon as the time column has some text contents, the
layouting engine would reserve only 7.199pt of height for the title of
some short schedule entries (even regardless of page height) for some
reason. Since the font size is set to 8, even the first "line" of the
schedule entry title text would then be completely omitted from rendering.

We fix the time column width for eternity by setting the width manually
and wrapping the contents in a position: absolute View.
The fixed width is calculated using a heuristic, based on the presumably
longest time displayed in the column. This heuristic could be greatly
improved if we used fontkit to actually measure the width of the string
as it will be rendered, but then we'd have to re-implement some of the
cascading styles logic ourselves... So for the moment, this simple
heuristic will have to do.

carlobeltrame committed:
Fix adaptively shrinking padding elements

carlobeltrame committed:
Stop treating first and last storyboard columns as rich text

carlobeltrame committed:
Avoid using empty View elements to display nothing

carlobeltrame committed:
Fix storyboard column widths

carlobeltrame committed:
Avoid empty View elements more thoroughly

carlobeltrame committed:
Implement ordered list enumeration

carlobeltrame committed:
Use the vuetify layout algorithm for placing schedule entries horizontally

carlobeltrame committed:
Workarounds for web worker issues with vite

carlobeltrame committed:
Give the document a more descriptive internal name

carlobeltrame committed:
Adjust default display of rendered pdf

pagedmode=none hides the bookmarks sidebar in firefox
zoom=page-fit is also a firefox / pdf.js feature
https://github.com/mozilla/pdf.js/wiki/Viewer-options#options-after-the-

navpanes=0 hides the bookmarks sidebar in chrome, but not firefox
https://stackoverflow.com/a/2105095

carlobeltrame committed:
Implement table of contents for the existing pdf components

carlobeltrame committed:
Refactor camp print document to delegate more logic to the actual components

carlobeltrame committed:
Implement cover page

carlobeltrame committed:
Extract category label into reusable component

carlobeltrame committed:
Implement story printing

carlobeltrame committed:
Merge remote-tracking branch 'origin/devel' into improve-react-print

usu commented:
Ready for review or you prefer to implement the remaining items first?

carlobeltrame commented:
Feel free to review. Later PRs can then be a little smaller again :)

usu commented on frontend/src/components/print/print-react/components/tableOfContents/entry/Program.jsx:
```diff
@@ -0,0 +1,45 @@
+  }
+  return periods.map((period) => {
+    const periodEntry = (
+      <Link style={tocStyles.entry} href={`#${props.id}-${period.id}`}>
```
This link currently jumps to the end of the period instead of to the beginning. Don't know why, though.

carlobeltrame replied:
Found the problem. It's because the target id is defined on a `<View>` which is split across multiple pages. Apparently each of these split instances gets a copy of the same id... Probably a bug. But I worked around it by adding an empty `<View>` at the top of the program and moved the id and bookmark onto there.

carlobeltrame committed:
Small fixes

carlobeltrame committed:
Fix program bookmark jumping to the last instead of the first page of the program

BacLuc commented on frontend/src/components/print/print-react/components/picasso/longestTime.js:
```diff
@@ -0,0 +1,9 @@
+import dayjs from '@/common/helpers/dayjs.js'
+
+export default function longestTime(times) {
```
Because the last time is the longest time?
This is different depending on the locale?

carlobeltrame replied:
Yes, depending on the locale, and in the future also depending on a dynamically adjusted times array (depending on the schedule entries in the period).

carlobeltrame committed:
Fix 'Each child in a list should have a unique "key" prop' warning

BacLuc commented on frontend/src/components/print/print-react/i18n.js:
```diff
@@ -1,8 +1,8 @@
 import {
-  translate,
+  compileToFunction,
```
what happened here?

carlobeltrame replied:
optimized imports in the IDE, which sorted them alphabetically

BacLuc reviewed (commented):
I have currently the error "document is not defined"

carlobeltrame committed:
Fix 'document is not defined'

This error happens, when Vue is imported in the web worker. Vue is not
designed for that, and will try to access the global document variable.
But that variable is not available inside a web worker. So we must avoid
importing Vue in the print scripts at all costs.

carlobeltrame commented:
> I have currently the error "document is not defined"

Should be resolved now. It was caused by the commons helper dateLong which imports Vue. Vue can not run inside a web worker, because it relies on accessing the global 'document' variable, which is not available inside web workers.

BacLuc reviewed (commented):
I have a warning:

Warning: Each child in a list should have a unique "key" prop. See https://fb.me/react-warning-keys for more information.
    in RichText (created by Storyboard)
    in VIEW (created by Storyboard)
    in VIEW (created by Storyboard)
    in VIEW (created by Storyboard)
    in Storyboard (created by ContentNode)
    in ContentNode (created by ColumnLayout)
    in VIEW (created by ColumnLayout)
    in VIEW (created by ColumnLayout)
    in ColumnLayout (created by ContentNode)
    in ContentNode (created by ColumnLayout)
    in VIEW (created by ColumnLayout)
    in VIEW (created by ColumnLayout)
    in ColumnLayout (created by ContentNode)
    in ContentNode (created by ScheduleEntry)
    in VIEW (created by ScheduleEntry)
    in ScheduleEntry (created by Program)
    in VIEW (created by Program)
    in PAGE (created by Program)
    in Program (created by PDFDocument)
    in DOCUMENT (created by PDFDocument)
    in PDFDocument [2bf3e71d.iife.js:26463:40](http://localhost:3000/node_modules/.vite/.bundle.iife/2bf3e71d.iife.js)
Warning: Each child in a list should have a unique "key" prop. See https://fb.me/react-warning-keys for more information.
    in RichText (created by SafetyConcept)
    in VIEW (created by SafetyConcept)
    in SafetyConcept (created by ContentNode)
    in ContentNode (created by ColumnLayout)
    in VIEW (created by ColumnLayout)
    in VIEW (created by ColumnLayout)
    in ColumnLayout (created by ContentNode)
    in ContentNode (created by ColumnLayout)
    in VIEW (created by ColumnLayout)
    in VIEW (created by ColumnLayout)
    in ColumnLayout (created by ContentNode)
    in ContentNode (created by ScheduleEntry)
    in VIEW (created by ScheduleEntry)
    in ScheduleEntry (created by Program)
    in VIEW (created by Program)
    in PAGE (created by Program)
    in Program (created by PDFDocument)
    in DOCUMENT (created by PDFDocument)
    in PDFDocument
    
Somewhere there is no key set, or the key is not unique

carlobeltrame commented:
> I have a warning:
> 
> Warning: Each child in a list should have a unique "key" prop. 
>
> Somewhere there is no key set, or the key is not unique

Yes. By now I fixed all of these warnings except for the RichText component, which still has some hidden "problem" somewhere. But since that message should only become a real problem in reactive applications, this warning can really be ignored in our case of static pdf rendering. So maybe a future improvement would be a custom warning handler which just silents these specific warnings.
